subsys/shell/modules/devmem_service.c

```diff
@@ -111,22 +112,22 @@ static int cmd_dump(const struct shell *sh, size_t argc, char **argv)
 	while ((rv = getopt(argc, argv, "a:s:w:")) != -1) {
 		switch (rv) {
 		case 'a':
-			addr = (mem_addr_t)strtoul(optarg, NULL, 16);
-			if (addr == 0 && errno == EINVAL) {
+			addr = (mem_addr_t)shell_strtoul(optarg, 16, &err);
+			if (err != 0) {
 				shell_error(sh, "invalid addr '%s'", optarg);
 				return -EINVAL;
 			}
 			break;
 		case 's':
-			size = (size_t)strtoul(optarg, NULL, 0);
-			if (size == 0 && errno == EINVAL) {
+			size = (size_t)shell_strtoul(optarg, 0, &err);
+			if (err != 0) {
 				shell_error(sh, "invalid size '%s'", optarg);
 				return -EINVAL;
 			}
 			break;
 		case 'w':
-			width = (size_t)strtoul(optarg, NULL, 0);
-			if (width == 0 && errno == EINVAL) {
```

Comment:
we do not need to validate that the `width` is != 0 ?

Reply:
I suppose that the original code is intended to only check the success of the `strtoul` operation, not the width itself, since its validity will be checked later in `memory_dump`.

---
By the way, according to https://pubs.opengroup.org/onlinepubs/9699919799.orig/functions/strtoul.html

1) If no conversion could be performed, **0** shall be returned and errno **may be set to** [EINVAL].
2) If the correct value is outside the range of representable values, {ULONG_MAX} or {ULLONG_MAX} shall be returned and errno set to [ERANGE].

It seems that the caller follows the first expectation, but since it states **may be** set, it ends up relying on the underlying `stdlib` implementation.